File: it-agents/skills/code-architect/references/code-examples/go/composite.go
// Composite: compose objects into tree structures and treat individual
// leaves and whole groups through one uniform interface.
package main

import "fmt"

// Component is the shared interface for leaves and containers.
type Component interface {
	Size() int
}

// Leaf is a primitive element with no children.
type Leaf struct {
	weight int
}

func (l Leaf) Size() int { return l.weight }

// Container holds children and aggregates their sizes recursively.
type Container struct {
	children []Component
}

func (c *Container) Add(child Component) {
	c.children = append(c.children, child)
}

func (c *Container) Size() int {
	total := 0
	for _, child := range c.children {
		total += child.Size()
	}
	return total
}

func main() {
	root := &Container{}
	root.Add(Leaf{weight: 3})

	sub := &Container{}
	sub.Add(Leaf{weight: 5})
	sub.Add(Leaf{weight: 2})
	root.Add(sub)

	fmt.Println("total size:", root.Size())
}

File: it-agents/skills/code-architect/references/code-examples/rust/composite.rs
// Composite — compose objects into tree structures and treat individual leaves
// and whole containers uniformly through one common interface.

trait Node {
    fn size(&self) -> u64;
    fn print(&self, indent: usize);
}

struct File {
    name: String,
    bytes: u64,
}

struct Folder {
    name: String,
    children: Vec<Box<dyn Node>>,
}

impl Node for File {
    fn size(&self) -> u64 {
        self.bytes
    }
    fn print(&self, indent: usize) {
        println!("{}- {} ({} B)", " ".repeat(indent), self.name, self.bytes);
    }
}

impl Node for Folder {
    fn size(&self) -> u64 {
        self.children.iter().map(|c| c.size()).sum()
    }
    fn print(&self, indent: usize) {
        println!("{}+ {}/ ({} B total)", " ".repeat(indent), self.name, self.size());
        for child in &self.children {
            child.print(indent + 2);
        }
    }
}

fn main() {
    let root = Folder {
        name: "root".into(),
        children: vec![
            Box::new(File { name: "readme.txt".into(), bytes: 120 }),
            Box::new(Folder {
                name: "src".into(),
                children: vec![
                    Box::new(File { name: "main.rs".into(), bytes: 800 }),
                    Box::new(File { name: "lib.rs".into(), bytes: 400 }),
                ],
            }),
        ],
    };

    root.print(0);
    println!("Grand total: {} bytes", root.size());
}
